### ISSUE TITLE
DisplayServer.window_set_mode has several issues [MacOS]

### ISSUE THREAD

**Issue by kkosovsky** [godotengine/godot#84142]
### Godot version

4.1.1.stable, 4.1.2.stable

### System information

MacOS Ventura - Apple M2 Max - Metal 3 - Compatibility Renderer

### Issue description

There are serious problems with changing window mode through `DisplayServer`'s `window_set_mode` function.
![image](https://github.com/godotengine/godot/assets/13417391/c00bde0f-5e2a-443f-927b-d95970030d0e)

https://github.com/godotengine/godot/assets/13417391/da8a0aaf-4970-477c-b304-484a52f28431

https://github.com/godotengine/godot/assets/13417391/9d0824f8-0a66-4317-9832-e2f6f4b8a19c

There are several cases of unexpected behavior. More in `steps to reproduce`.

### Steps to reproduce

Test 1:
Window -> Maximized -> Window
`CORRECT`

Test 2:
Window -> Maximized -> Fullscreen
The window behaves oddly between Maxmized -> Fullscreen mode change. It looks like it's going to window mode between maximized and fullscreen.
`INCORRECT`

Test 3:
Window -> Maximized -> Exclusive Fullscreen
Same as Test 2.
`INCORRECT`

Test 3:
Window -> Maximized -> Fullscreen -> Window
Window has correct size. 
`CORRECT`

Test 4:
Window -> Fullscreen -> Exclusive Fullscreen
Instead of going from Fullscreen to Exclusive Fullscreen it goes back to Window.
`INCORRECT`

Test 5:
Window -> Fullscreen -> Exclusive Fullscreen -> Window
Then after changing to Window, it goes to fullscreen
`INCORRECT`

Test 6:
Window -> Fullscreen -> Exclusive Fullscreen -> Window -> Exclusive Fullscreen
Then after changing to Fullscreen it goes back to window.
`INCORRECT`

### Minimal reproduction project

[window_mode_issue.zip](https://github.com/godotengine/godot/files/13198364/window_mode_issue.zip)

**Comment by bruvzg** [comment]
> The window behaves oddly between Maxmized -> Fullscreen mode change. It looks like it's going to window mode between maximized and fullscreen.

That's expected, otherwise it won't be able to restore window size when exiting full-screen.

**Comment by kkosovsky** [comment]
> > The window behaves oddly between Maxmized -> Fullscreen mode change. It looks like it's going to window mode between maximized and fullscreen.
> 
> That's expected, otherwise it won't be able to restore window size when exiting full-screen.

Yes, but it should go to small window size and then to maximized. There shouldn't be an in between size, as I imagine it.

**Comment by Cyangmou** [comment]
> > The window behaves oddly between Maxmized -> Fullscreen mode change. It looks like it's going to window mode between maximized and fullscreen.
> 
> That's expected, otherwise it won't be able to restore window size when exiting full-screen.

It's also not restoring window size on WINDOWS when you selected maximized before.

I mean Window logically should be a small window.
Window_Maximized should be the maximized window.

**Comment by jitspoe** [comment]
I've fixed the maximized->fullscreen->windowed mode bug specifically for Windows here: https://github.com/godotengine/godot/pull/84504/files

Pretty trivial fix.  I'm assuming the same logic error exists for Mac as well (`else if (p_mode == WINDOW_MODE_WINDOWED)` just needs the `else` removed), but I don't have a mac to build and test on.  If any mac devs want to make a contribution, this should be an easy one! :)

**Comment by bruvzg** [comment]
> I'm assuming the same logic error exists for Mac as well (else if (p_mode == WINDOW_MODE_WINDOWED) just needs the else removed)

This will break switching `maximize` → `fullscreen` → `window` switch (macOS do not have real maximized state, so exiting full-screen will return it to previous window size).
